src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala

```diff
@@ -0,0 +1,167 @@
+
+  def newOptions = new Options
+
+  def supportsMultiallelic = false
+
+  def requiresVDS = true
+
+  def run(state: State, options: Options): State = {
+    val vds = state.vds
+
+    val pathVA = Parser.parseAnnotationRoot(options.rootVA, Annotation.VARIANT_HEAD)
+    val pathGA = Parser.parseAnnotationRoot(options.rootGA, Annotation.GLOBAL_HEAD) // FIXME: included only to check for errors
+
+    val symTab = Map(
+      "s" -> (0, TSample),
+      "sa" -> (1, vds.saSignature))
+
+    val ec = EvalContext(symTab)
+
+    def toDouble(t: BaseType, code: String): Any => Double = t match {
+      case TInt => _.asInstanceOf[Int].toDouble
+      case TLong => _.asInstanceOf[Long].toDouble
+      case TFloat => _.asInstanceOf[Float].toDouble
+      case TDouble => _.asInstanceOf[Double]
+      case TBoolean => _.asInstanceOf[Boolean].toDouble
+      case _ => fatal(s"Sample annotation `$code' must be numeric or Boolean, got $t")
+    }
+
+    val (yT, yQ) = Parser.parse(options.ySA, ec)
+    val yToDouble = toDouble(yT, options.ySA)
+    val ySA = vds.sampleIdsAndAnnotations.map { case (s, sa) =>
+      ec.setAll(s, sa)
+      yQ().map(yToDouble)
+    }
+
+    val (covT, covQ) = Parser.parseExprs(options.covSA, ec).unzip
+    val covToDouble = (covT, options.covSA.split(",").map(_.trim)).zipped.map(toDouble)
+    val covSA = vds.sampleIdsAndAnnotations.map { case (s, sa) =>
+      ec.setAll(s, sa)
+      (covQ.map(_ ()), covToDouble).zipped.map(_.map(_))
+    }
+
+    val (yForCompleteSamples, covForCompleteSamples, completeSamples) =
+      (ySA, covSA, vds.sampleIds)
+        .zipped
+        .filter((y, c, s) => y.isDefined && c.forall(_.isDefined))
+
+    val yArray = yForCompleteSamples.map(_.get).toArray
+    val y = DenseVector(yArray)
+
+    val covArray = covForCompleteSamples.flatMap(_.map(_.get)).toArray
+    val k = covT.size
+    val cov =
+      if (k == 0)
+        None
+      else
+        Some(new DenseMatrix(
+          rows = completeSamples.size,
+          cols = k,
+          data = covArray,
+          offset = 0,
+          majorStride = k,
+          isTranspose = true))
+
+    val n = y.size
+    val d = n - k - 2
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${plural(k, "covariate")} with intercept implies $d degrees of freedom.")
+
+    info(s"lmmreg: running lmmreg on $n samples with $k sample ${plural(k, "covariate")}...")
+
+    val C: DenseMatrix[Double] = cov match {
+      case Some(dm) => DenseMatrix.horzcat(DenseMatrix.ones[Double](n, 1), dm)
+      case None => DenseMatrix.ones[Double](n, 1)
+    }
+
+    val optDelta = Option(options.delta).map(_.doubleValue())
+    optDelta.foreach(delta =>
+      if (delta <= 0d)
+          fatal(s"delta must be positive, got ${ delta }"))
+
+    // FIXME: is this right? and can comma's occur inside backticks?
+    val covNames = options.covSA.trim() match {
+      case "" => List("intercept")
+      case names => "intercept" +: names.split(",").map(_.trim()).toList
```

Comment:
I think this is correct because all the weird cases should trigger errors in the parser before we get here. It still makes me uncomfortable because there's this implicit dependency that the parser is working the same way this code does.